presto-spi/src/main/java/io/prestosql/spi/block/Int128ArrayBlock.java

```diff
@@ -0,0 +1,234 @@
+            consumer.accept(valueIsNull, sizeOf(valueIsNull));
+        }
+        consumer.accept(this, (long) INSTANCE_SIZE);
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positionCount;
+    }
+
+    @Override
+    public long getLong(int position, int offset)
+    {
+        checkReadablePosition(position);
+        if (offset == 0) {
+            return values[(position + arrayOffset) * 2];
+        }
+        if (offset == 8) {
+            return values[((position + arrayOffset) * 2) + 1];
+        }
+        throw new IllegalArgumentException("offset must be 0 or 8");
+    }
+
+    @Override
+    public boolean mayHaveNull()
+    {
+        return valueIsNull != null;
+    }
+
+    @Override
+    public boolean isNull(int position)
+    {
+        checkReadablePosition(position);
+        return valueIsNull != null && valueIsNull[position + arrayOffset];
+    }
+
+    @Override
+    public void writePositionTo(int position, BlockBuilder blockBuilder)
+    {
+        checkReadablePosition(position);
+        blockBuilder.writeLong(values[(position + arrayOffset) * 2]);
+        blockBuilder.writeLong(values[((position + arrayOffset) * 2) + 1]);
+        blockBuilder.closeEntry();
+    }
+
+    @Override
+    public Block getSingleValueBlock(int position)
+    {
+        checkReadablePosition(position);
+        return new Int128ArrayBlock(
+                0,
+                1,
+                isNull(position) ? new boolean[] {true} : null,
+                new long[] {
+                        values[(position + arrayOffset) * 2],
+                        values[((position + arrayOffset) * 2) + 1]});
+    }
+
+    @Override
+    public Block copyPositions(int[] positions, int offset, int length)
+    {
+        checkArrayRange(positions, offset, length);
+
+        boolean[] newValueIsNull = null;
+        if (valueIsNull != null) {
+            newValueIsNull = new boolean[length];
+        }
+        long[] newValues = new long[length * 2];
+        for (int i = 0; i < length; i++) {
+            int position = positions[offset + i];
+            checkReadablePosition(position);
+            if (valueIsNull != null) {
+                newValueIsNull[i] = valueIsNull[position + arrayOffset];
+            }
+            newValues[i * 2] = values[(position + arrayOffset) * 2];
+            newValues[(i * 2) + 1] = values[((position + arrayOffset) * 2) + 1];
+        }
+        return new Int128ArrayBlock(0, length, newValueIsNull, newValues);
+    }
+
+    @Override
+    public Block getRegion(int positionOffset, int length)
+    {
+        checkValidRegion(getPositionCount(), positionOffset, length);
+
+        return new Int128ArrayBlock(positionOffset + arrayOffset, length, valueIsNull, values);
+    }
+
+    @Override
+    public Block copyRegion(int positionOffset, int length)
+    {
+        checkValidRegion(getPositionCount(), positionOffset, length);
+
+        positionOffset += arrayOffset;
```

Comment:
you probably shouldn't multiply `arrayOffset` * 2, below you have code:
```
positionOffset * 2
```
which resolves to: `(positionOffset + arrayOffset) * 2`
but I think it should be
```
arrayOffset + positionOffset * 2
```